tls/s2n_ktls_io.c

```diff
@@ -0,0 +1,239 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <sys/socket.h>
+
+#include "tls/s2n_ktls.h"
+#include "utils/s2n_socket.h"
+
+/*
+ * sendmsg and recvmsg are syscalls which can be used to send 'real' data along
+ * with ancillary data. Ancillary data is used to communicate to the socket the
+ * type of the TLS record being sent/received.
+ *
+ * Ancillary data macros (CMSG_*) are platform specific and gated.
+ */
+
+S2N_RESULT s2n_ktls_send_msg_impl(int sock, struct msghdr *msg,
+        struct iovec *msg_iov, s2n_blocked_status *blocked, ssize_t *result)
+{
+    RESULT_ENSURE_REF(msg);
+    RESULT_ENSURE_REF(msg_iov);
+    RESULT_ENSURE_REF(blocked);
+    RESULT_ENSURE_REF(result);
+
+    /* set send buffer */
+    msg->msg_iov = (struct iovec *) msg_iov;
+    msg->msg_iovlen = 1;
```

Comment:
Realized that I should be taking a `size_t count` arg to set this since we also support the `s2n_sendv` and the application could pass a more than 1 iovec.